codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithygo/codegen/knowledge/GoPointableIndex.java

```diff
@@ -76,6 +73,7 @@ public class GoPointableIndex implements KnowledgeIndex {
             // Interfaces
```

Comment:
Changing PointableIndex can break other cases. Are we sure these changes are safe?

Reply:
Pointer vs non pointer issue has been biting us a lot. I think rather then doing this as a part of this PR we need to take a closer look into it in its own sub task. So, I reverted few. Things that are changed and are not reverted are:
- If a `membershape` has required trait it is always non pointer. We talked about this way back has this changed?
- I renamed `NUMBER_SHAPE` variable to `KNOWN_POINTER_TYPE` because if we find some other shape with similar issue we can add it to this variable
- String seems to be always considered a pointer type. `DafnyToSmithyShapeVisitor` does not handle a case where string is non pointer (See the function return type). I have made a comment in a sim so that we can dive deep into it in a separate sub task that is only related to Pointer

Reply:
The reason being legacy Box / Default traits in smithy. So there's no one line answer and this needs to be handled on case by case basis.

1. Yes - this is not always true, more so when dealing with existing aws-models and go sdk.
2. Makes sense.
3. Again this goes back to 1 but yeah I'm yet to see a case in aws-sdk where string is non-pointer type.